### DISCUSSION TITLE
Route caching causes web server exception

### DISCUSSION THREAD

**Discussion by acharseth** [deployphp/deployer#3673] [Help needed]
I am not sure if this issue belongs in this forum but I try.

I have configured `deploy.php` without modifying the caching part. Now when I let deploy.php run `artisan:route:cache` it causes the following error message when I try to reach my web application in the browser with ULR: https://foo/bar/current/public/
```
The GET method is not supported for route /. Supported methods: HEAD
```
If I then run `php artisan route:clear`, or override this in `deploy.php`, this error disappears. I guess it has a performance penalty though. If I then run `php artisan route:cache` the error reappears.

If I instead use the URL: https://foo/bar/current/public/dashboard it works.

In `web.php` I have a default root route like this:
```php
Route::get('/', function () {
    return view('welcome');
});
```

If I remember correct the .htaccess file is untouched with the following standard RewriteCond and ...Rule:
```
    # Redirect Trailing Slashes If Not A Folder...
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_URI} (.+)/$
    RewriteRule ^ %1 [L,R=301]

    # Send Requests To Front Controller...
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME} !-f
    RewriteRule ^ index.php [L]
```

How can I make the route cache work?

**Comment by antonmedv** [comment]
1. **Closure-Based Routes**: Laravel's route cache doesn't support closure-based routes. You've mentioned you have a closure for your root route in `web.php`. This can be a possible reason for the problem. Convert your closure route to use a controller method:

   ```php
   // Before
   Route::get('/', function () {
       return view('welcome');
   });
   
   // After
   Route::get('/', 'HomeController@index');
   ```

   And in `HomeController`:

   ```php
   public function index() {
       return view('welcome');
   }
   ```

2. **Check the `.env` File**: Ensure that your `.env` file (and its cached version) doesn't have any inconsistencies regarding the app URL or any other configurations. Any changes in the `.env` file should be followed by:

   ```bash
   php artisan config:clear
   ```

   Then try caching the routes again.

3. **Review Your Deployment Script**: Ensure that the order of artisan commands in your deployment script is correct. Generally, you might want to clear the config and route cache before regenerating them.

   ```bash
   php artisan config:clear
   php artisan route:clear
   php artisan route:cache
   ```

4. **.htaccess File**: The `.htaccess` file seems correct, but make sure it's in the correct directory (`public`) and that the web server has the necessary permissions to read it. Also, ensure your server is set up to respect `.htaccess` configurations.

5. **Middleware & Route Grouping**: Ensure you don't have any middleware or route group configurations that might interfere with the root route. For instance, there might be a middleware or group that forces a trailing slash, and the behavior with caching may differ.

6. **Check for Duplicate Routes**: Ensure that there aren't any duplicate routes defined. Duplicate routes can cause issues when caching, as Laravel might cache the wrong route.

7. **Logs & Debugging**: Look at Laravel's logs (`storage/logs/laravel.log`) and the web server's logs to see if there's any additional information about the issue. This can provide insights into any underlying problems.
